### PR TITLE
feat: Add support for Polish shippers + devcontainer

### LINKED ISSUE #595: [Feature Request] GLS
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
How hard it it to add support for GLS?

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by kubawolanin**
## Proposed change

image

Add Polish postal services to Mail And Packages custom component. The following services are supported:
- Poczta Polska SA
- InPost.pl
- DPD Polska
- GLS
- DHL (updates existing)
- Amazon.pl (updates existing)

At the same time I've adjusted unit tests to maintain proper coverage.
image

Also:
1. Adds devcontainer for easier development
2. Refactors splitting shipper's keys by `_` so that it supports multiple underscore chars in shipper ID (like `poczta_polska_delivering` etc).

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests
- [ ] Documentation update
- [x] Adds a new shipper
- [x] Update existing shipper

## Additional information

- This PR fixes or closes issue: fixes #595 (only for Poland)
- This PR is related to issue:

### CODE REVIEW

File: custom_components/mail_and_packages/helpers.py [hunk 1]
```diff
@@ -380,12 +380,12 @@ def fetch(
...
         count[AMAZON_EXCEPTION_ORDER] = info[ATTR_ORDER]
     elif "_packages" in sensor:
-        prefix = sensor.split("_")[0]
+        prefix = sensor.replace("_packages", "")
         delivering = fetch(hass, config, account, data, f"{prefix}_delivering")
         delivered = fetch(hass, config, account, data, f"{prefix}_delivered")
         count[sensor] = delivering + delivered
     elif "_delivering" in sensor:
-        prefix = sensor.split("_")[0]
+        prefix = sensor.replace("_delivering", "")
```
Comment (kubawolanin) [inline custom_components/mail_and_packages/helpers.py:388]: This is to support sensor ids like `inpost_pl_delivering` and so on

File: custom_components/mail_and_packages/helpers.py [hunk 2]
```diff
@@ -857,8 +858,13 @@ def get_count(
             )
             found.append(data[0])
 
-    if ATTR_PATTERN in SENSOR_DATA[f"{sensor_type.split('_')[0]}_tracking"].keys():
-        track = SENSOR_DATA[f"{sensor_type.split('_')[0]}_tracking"][ATTR_PATTERN][0]
+    if (
+        ATTR_PATTERN
+        in SENSOR_DATA[f"{'_'.join(sensor_type.split('_')[:-1])}_tracking"].keys()
+    ):
+        track = SENSOR_DATA[f"{'_'.join(sensor_type.split('_')[:-1])}_tracking"][
```
Comment (kubawolanin) [inline custom_components/mail_and_packages/helpers.py:865]: This is to support sensor ids like `inpost_pl_tracking` and so on

File: custom_components/mail_and_packages/const.py [hunk 3]
```diff
@@ -202,6 +237,114 @@
     },
     "royal_packages": {},
     "royal_tracking": {"pattern": ["[A-Za-z]{2}[0-9]{9}GB"]},
+    # Poczta Polska SA
+    "poczta_polska_delivered": {},
```
Comment (firstof9) [inline custom_components/mail_and_packages/const.py:241]: I'd like to try to keep these names simple like `pocztapolska_delivered` and `inpostpl_delivered`

File: tests/test_helpers.py [hunk 4]
```diff
@@ -1042,28 +1042,21 @@ async def test_image_file_name(
...
+    assert result["count"] == 35
 
     result = amazon_exception(mock_imap_amazon_exception, ["testemail@fakedomain.com"])
-    assert result["count"] == 7
+    assert result["count"] == 36
     assert (
-        "Amazon domains to be checked: ['amazon.com', 'amazon.ca', 'amazon.co.uk', 'amazon.in', 'amazon.de', 'amazon.it', 'testemail@fakedomain.com']"
+        "Amazon domains to be checked: ['amazon.com', 'amazon.ca', 'amazon.co.uk', 'amazon.in', 'amazon.de', 'amazon.it', 'amazon.pl', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'testemail@fakedomain.com']"
```
Comment (firstof9) [inline tests/test_helpers.py:1051]:
```suggestion
        "Amazon domains to be checked: ['amazon.com', 'amazon.ca', 'amazon.co.uk', 'amazon.in', 'amazon.de', 'amazon.it', 'amazon.pl', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'testemail@fakedomain.com']"
```

**Review by firstof9 — CHANGES_REQUESTED** [review]

File: tests/test_helpers.py [hunk 5]
```diff
@@ -800,7 +800,7 @@ async def test_amazon_shipped_order_it_count(hass, mock_imap_amazon_shipped_it):
     with patch("datetime.date") as mock_date:
         mock_date.today.return_value = date(2021, 12, 1)
         result = get_items(mock_imap_amazon_shipped_it, "count")
-        assert result == 1
+        assert result == 0
```
Comment (firstof9) [inline tests/test_helpers.py:803]:
```suggestion
        assert result == 1
```

File: tests/test_helpers.py [hunk 6]
```diff
@@ -1042,28 +1042,21 @@ async def test_image_file_name(
...
+        "Amazon domains to be checked: ['amazon.com', 'amazon.ca', 'amazon.co.uk', 'amazon.in', 'amazon.de', 'amazon.it', 'amazon.pl', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'fakeuser@fake.email', 'fakeuser2@fake.email', 'testemail@fakedomain.com']"
         in caplog.text
     )
 
 
 async def test_hash_file():
     """Test file hashing function."""
     result = hash_file("tests/test_emails/amazon_delivered.eml")
-    assert result == "7f9d94e97bb4fc870d2d2b3aeae0c428ebed31dc"
+    assert result == "0e2c7e290472fffc0cb1c6b6b860b4bd7f386f7c"
```
Comment (firstof9) [inline tests/test_helpers.py:1059]:
```suggestion
    assert result == "7f9d94e97bb4fc870d2d2b3aeae0c428ebed31dc"
```

Comment (firstof9) [reply tests/test_helpers.py:1059]: This shouldn't change.

File: tests/test_helpers.py (hunk 5 above)
Comment (firstof9) [reply tests/test_helpers.py:803]: This should always be `1`.

**Comment by firstof9** [comment]
> Currently there's a bug in this PR breaking whenever it _encounters non-ascii characters in the emails subject_

You can try adjusting:
https://github.com/moralmunky/Home-Assistant-Mail-And-Packages/blob/24ac9a18e5cfc66781d82843690b5a7f28e2a253/custom_components/mail_and_packages/helpers.py#L519

to
```python
value = account.search(None, imap_search.encode('utf-8'))
```

File: custom_components/mail_and_packages/const.py (hunk 3 above)
Comment (kubawolanin) [reply custom_components/mail_and_packages/const.py:241]: The issue with this is that these IDs are generated from names in `SensorEntityDescription`. 
`Mail InPost.pl Delivered` then becomes `mail_inpost_pl_delivered` and so on.

I suspect there will be more cases like this in the future, where shipper's name contain multiple words.

File: tests/test_helpers.py [hunk 7]
```diff
@@ -1042,20 +1042,13 @@ async def test_image_file_name(
...
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-    ]
-    assert result["count"] == 6
+    assert result["order"] == ["123-1234567-1234567"] * 35
+    assert result["count"] == 35
 
     result = amazon_exception(mock_imap_amazon_exception, ["testemail@fakedomain.com"])
-    assert result["count"] == 7
+    assert result["count"] == 36
```
Comment (firstof9) [inline tests/test_helpers.py:1049]:
```suggestion
    assert result["count"] == 8
```
This should be the amount of domains in the `AMAZON_DOMAINS` const plus the 1 extra added domain in the call.

File: tests/test_helpers.py [hunk 8]
```diff
@@ -1042,20 +1042,13 @@ async def test_image_file_name(
...
-    assert result["order"] == [
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-        "123-1234567-1234567",
-    ]
-    assert result["count"] == 6
+    assert result["order"] == ["123-1234567-1234567"] * 35
```
Comment (firstof9) [inline tests/test_helpers.py:1045]:
```suggestion
    assert result["order"] == ["123-1234567-1234567"] * 7
```
This should be the amount of domains in the `AMAZON_DOMAINS` const.

File: tests/test_helpers.py [hunk 9]
```diff
@@ -1042,20 +1042,13 @@ async def test_image_file_name(
...
+    assert result["count"] == 35
```
Comment (firstof9) [inline tests/test_helpers.py:1046]:
```suggestion
    assert result["count"] == 7
```
This should be the amount of domains in the `AMAZON_DOMAINS` const.

**Review by firstof9 — CHANGES_REQUESTED** [review]

**Comment by kubawolanin** [comment]
@firstof9 apologies for my absence here, have some personal stuff to take care of. 
Thanks for rebasing this branch with latest dev!

**Comment by firstof9** [comment]
No worries, life takes priority.

**Comment by firstof9** [comment]
It's possible we might need to break this PR up into 3 or 4 smaller PRs.

**Comment by firstof9** [comment]
Actually I think I see why it's doing what it's doing now.

**Comment by firstof9** [comment]
Ok looks like the final piece is just to run black against `custom_components` via `black custom_components/` and the PR should be sorted.

**Review by firstof9 — APPROVED** [review]

**Comment by firstof9** [comment]
Thanks for the contribution!

**Pull request merged**